src/databricks/sql/auth/oauth.py

```diff
@@ -0,0 +1,286 @@
+
+import click
+
+import jwt
+from jwt import PyJWTError
+
+import oauthlib.oauth2
+from oauthlib.oauth2.rfc6749.errors import OAuth2Error
+
+import requests
+from requests.exceptions import RequestException
+
+from databricks_cli.utils import error_and_quit
+
+try:
+    from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+except ImportError:
+    from http.server import BaseHTTPRequestHandler, HTTPServer
+
+
+# This could use 'import secrets' in Python 3
+def token_urlsafe(nbytes=32):
+    tok = os.urandom(nbytes)
+    return base64.urlsafe_b64encode(tok).rstrip(b'=').decode('ascii')
+
+
+# This could be datetime.timezone.utc in Python 3
+class UTCTimeZone(tzinfo):
+    """UTC"""
+    def utcoffset(self, dt):
+        #pylint: disable=unused-argument
+        return timedelta(0)
+
+    def tzname(self, dt):
+        #pylint: disable=unused-argument
+        return "UTC"
+
+    def dst(self, dt):
+        #pylint: disable=unused-argument
+        return timedelta(0)
+
+
+# Some constant values
+OIDC_REDIRECTOR_PATH = "oidc"
+REDIRECT_PORT = 8020
+UTC = UTCTimeZone()
+
+
+def get_client(client_id):
+    return oauthlib.oauth2.WebApplicationClient(client_id)
+
+
+def get_redirect_url(port=REDIRECT_PORT):
+    return "http://localhost:{port}".format(port=port)
+
+
+def fetch_well_known_config(idp_url):
+    known_config_url = "{idp_url}/.well-known/oauth-authorization-server".format(idp_url=idp_url)
+    try:
+        response = requests.request(method="GET", url=known_config_url)
+    except RequestException:
+        error_and_quit("Unable to fetch OAuth configuration from {idp_url}.\n"
+                       "Verify it is a valid workspace URL and that OAuth is "
+                       "enabled on this account.".format(idp_url=idp_url))
+
+    if response.status_code != 200:
+        error_and_quit("Received status {status} OAuth configuration from "
+                       "{idp_url}.\n Verify it is a valid workspace URL and "
+                       "that OAuth is enabled on this account."
+                       .format(status=response.status_code, idp_url=idp_url))
+    try:
+        return json.loads(response.text)
+    except json.decoder.JSONDecodeError:
+        error_and_quit("Unable to decode OAuth configuration from {idp_url}.\n"
+                       "Verify it is a valid workspace URL and that OAuth is "
+                       "enabled on this account.".format(idp_url=idp_url))
+
+
+def get_idp_url(host):
+    maybe_scheme = "https://" if not host.startswith("https://") else ""
+    maybe_trailing_slash = "/" if not host.endswith("/") else ""
+    return "{scheme}{host}{trailing}{path}".format(
+        scheme=maybe_scheme, host=host, trailing=maybe_trailing_slash, path=OIDC_REDIRECTOR_PATH)
+
+
+def get_challenge(verifier_string=token_urlsafe(32)):
+    digest = hashlib.sha256(verifier_string.encode('UTF-8')).digest()
+    challenge_string = base64.urlsafe_b64encode(digest).decode("UTF-8").replace('=', '')
```

Comment:
nit: inconsistent use of single and double quotes in the same method.

Reply:
addressed.